numpy/ma/core.py

```diff
@@ -6779,18 +6778,19 @@ def where(condition, x=_NoValue, y=_NoValue):
         ndtype = np.find_common_type([xv.dtype, yv.dtype], [])
 
     # Construct an empty array and fill it
-    d = np.empty(fc.shape, dtype=ndtype).view(MaskedArray)
-    np.copyto(d._data, xv.astype(ndtype), where=fc)
-    np.copyto(d._data, yv.astype(ndtype), where=notfc)
+    data = np.where(fc, xv.astype(ndtype), yv.astype(ndtype))
+    d = data.view(MaskedArray)
 
     # Create an empty mask and fill it
-    mask = np.zeros(fc.shape, dtype=MaskType)
-    np.copyto(mask, getmask(x), where=fc)
-    np.copyto(mask, getmask(y), where=notfc)
-    mask |= getmaskarray(condition)
+    mask = np.where(fc, getmaskarray(x), getmaskarray(y))
+    np.copyto(mask, True, where=getmaskarray(condition))
 
+    if isinstance(mask.dtype.type, np.void):
+        needmask = np.any(np.ones(1, mask.dtype) == mask)
```

Comment:
I think this would mean that if there is a record in which only one element is masked, needmask would be `False`. If so, you'd want
```
needmask = not np.all(np.zeros(1, mask.dtype) == mask)
```